keps/sig-node/1977-container-notifier/README.md

```diff
@@ -0,0 +1,773 @@
+
+In phase 2, this API proposal adds an inline pod definition for container signals and allows the API object to send a request to trigger those signals. Logic in the container notifier controller will be moved to Kubelet. Kubelet watches the objects and runs the command when it is created and updates its status accordingly.
+
+In phase 3, a Probe may be added if needed as an inline pod definition to verify whether the signal is delivered or whether the command is run and results in the desired outcome.
+
+### API Changes
+
+#### Phase 1
+
+There are two parts in the API changes:
+* Add an inline pod definition for ContainerNotifierAction - define a command to run in the container.
+* Add a Notification API object - this is the request to trigger a container notifier.
+
+#### Phase 2
+
+* Add a signal to the inline pod definition:
+  * In the inline pod definition for ContainerNotifierAction, allow a signal to be defined and sent to the container.
+
+#### Phase 3
+
+* If needed, add an inline pod definition for Probe - probes the container to verify that ContainerNotifierAction has brought the container to the desired state.
+
+#### Phase 1 API Changes
+
+##### Inline Pod Definition for ContainerNotifier
+
+Add []ContainerNotifier to the Container struct:
+
+```
+type Container struct {
+     ......
+     // +optional
+     Lifecycle *Lifecycle `json:"lifecycle,omitempty" protobuf:"bytes,12,opt,name=lifecycle"`
+
+    // Notifiers can be triggered by Notification resources. Each notifier must
+    // have a unique name within the container definition, but the same name may be
+    // used in different containers of the same pod. When a Notification is created
+    // which selects a pod and matches the Notifier name, the Notifier will run in the
+    // container.
+    // +optional
+    Notifiers []ContainerNotifier
+    ......  
+}
+
+type ContainerNotifier struct {
+   // Name must be unique within a container.
+   // Names are label-style, with an optional prefix. Names without the prefix
+   // such as `quiesce` are reserved for Kubernetes-defined "well-known" names.
+   // Names with a prefix such as `example.com/label-style` are custom names.
+   Name string
+
+   // Action defines what happens when a notifier is triggered.  Exactly-once
+   // execution is not guaranteed, so this must be idempotent. When the action is
+   // triggered multiple times, it should achieve the same results. For example,
+   // when a quiesce command is run on a database, the database is quiesced. If the
+   // same quiesce command is run again, it must not have additional impact,
+   // because the database is already quiesced.
+   Action *ContainerNotifierAction
+
+    // Number of seconds after which the notifier times out.
+    // Defaults to 1 second. Minimum value is 1.
+    // +optional
+    TimeoutSeconds int32
+}
+
+// ContainerNotifier describes a command to run in the container.
+type ContainerNotifierAction struct {
```

Comment:
I was proposing that we get rid of this layer of nesting:

```
Pod {
  Spec: {
    Containers: [{
      Notifiers: [{
        Name: foo
        TimeoutSeconds: 10
        Action: {   # note no "handler"
          Exec: {
            Command: bar
          }
        }
      }]
    }]
  }
}
```

Reply:
yes, agreed

Reply:
+1

Reply:
ping